### ISSUE TITLE
model.py

### ISSUE THREAD

**Issue by GreenGrass31** [GreenGrass31/Game-Development-#1]
!pip install torchsummary
!pip install torch

import torch
import torch.nn as nn
import torchvision
from torchviz import make_dot
from torchsummary import summary

class convmodule(nn.Module):    #여러 개의 conv layer을 포함하는 모듈
    def __init__(self, in_channels, out_channels, **kwargs):
        super().__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, bias = False, **kwargs)    #해당 모듈을 구성할 conv layer의 형식 정의
        self.batchnorm = nn.BatchNorm2d(out_channels)    #해당 모듈에서 사용될 batchnorm layer 정의
        self.activation = nn.LeakyReLU(0.1)     
    def forward(self, x):
        return self.activation(self.batchnorm(self.conv(x)))    #this method is run when an instance is called
  
class yolo(nn.Module):
    def __init__(self, in_channels=3, **kwargs):    #depth = 3, 그 외 keyword argument 받음
        super().__init__()
        self.model_config = model_config
        self.in_channels = in_channels
        self.backbone = self.create_conv_layers(model_config)
        self.fully_connected = self.create_fc(**kwargs)

    def forward(self, x):
        x = self.backbone(x)
        return self.fully_connected(torch.flatten(x, start_dim = 1))  #start_dim = 1 to retain batch size 예를 들어 x = (64, 7, 7, 1024) 라면 순서대로 0, 1, 2, 3의 dim을 가짐. dim = 0이 batch size 이므로  

    def create_conv_layers(self, config):
        layers = []
        in_channels = self.in_channels    
        '''
        만약 model = yolo()라고 인스턴스 선언을 해줬다 하자. 이 인스턴스에는 in_channels라는 변수?가 있을거임. 근데 이 변수에다가 입력으로 들어온 in_channels에 해당하는 값을 넣어줌. 
        그리고 이 in_channels라는 값은 입력값의 레이블? 키워드의 느낌이라고 생각하면 됨. 
        인스턴스의 in_channels에 접근하기 위해서는 model.in_channels 경로로 접근해야 함. (init에서 self를 쓰는 게 이런 의미이다.)
        그리고 in_channels라는 입력값은 클래스 내 전역변수가 아니므로 클래스 내 다른 메서드에서 사용하기 위해서는 self.in_channels로 접근해야 함. 
        이 구체적인 경우에는 model.in_channel 
        '''
        
        for i in config:
            if i == "M":      #Maxpool
                layers += [nn.MaxPool2d(kernel_size=2, stride=2)]
            elif type(i) == tuple:       #conv layer
                layers += [convmodule(in_channels, i[1], kernel_size = i[0], stride = i[2], padding = i[3])]
                in_channels = i[1]
            elif type(i) == list:        #multiple conv layers
                layer1 = i[0]
                layer2 = i[1]
                num = i[2]
                for j in range(num):
                    layers += [convmodule(in_channels, layer1[1], kernel_size = layer1[0], stride = layer1[2], padding = layer1[3]), convmodule(in_channels, layer2[1], kernel_size = layer2[0], stride = layer2[2], padding = layer2[3])]
                    in_channels = layer2[1]
        return nn.Sequential(*layers)    #the asterisk is an unpacking operator which unpacks the list and passes each element as a seperate argument

    def create_fc(self, grid_size, num_box, num_class):
        return nn.Sequential(nn.Linear(1024 * grid_size * grid_size, 4096), nn.LeakyReLU(0.1), nn.Linear(4096, (grid_size * grid_size * (5 * num_box + num_class))))    #This FC layer results in a very large number of parameters. So tweak the size of the layer if necessary

model = yolo(grid_size = 7, num_box = 3, num_class = 20)
print(model)
